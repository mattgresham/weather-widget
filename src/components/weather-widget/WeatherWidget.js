// @flow
import React, { Component } from 'react';
import type { Units } from '../../types/units';
import style from './WeatherWidget.module.css';

type Props = {
  getWeather: Function,
  icon?: string,
  isWindShown: boolean,
  location?: string,
  temperature?: number,
  title?: string,
  units?: Units,
  wind?: string
};

class WeatherWidget extends Component<Props> {
  static defaultProps = {
    icon: '',
    location: '—',
    temperature: 0,
    title: '—',
    units: 'metric',
    wind: '—'
  };

  constructor(props: Props) {
    super(props);
    props.getWeather();
  }

  componentDidUpdate(prevProps: Props) {
    if (prevProps.units !== this.props.units) {
      this.props.getWeather();
    }
  }

  render() {
    const {
      icon,
      isWindShown,
      location,
      temperature,
      title,
      units,
      wind
    } = this.props;

    return (
      <div className={style.widget}>
        <div className={style.title}>{title || 'Title of widget'}</div>
        <div className={style.container}>
          <div className={style.iconContainer}>
            {icon && (
              <img
                alt=""
                className={style.icon}
                src={`https://openweathermap.org/img/w/${icon}.png`}
              />
            )}
          </div>
          <div className={style.weatherContainer}>
            <div>{location || '-'}</div>
            <div className={style.temperature}>
              {temperature || '-'}
              <span className={style.degree}>&deg;</span>
            </div>
            {isWindShown && (
              <div className={style.windContainer}>
                <span className={style.windLabel}>Wind</span> {wind || '-'}
                {units === 'metric' ? 'km/h' : 'mph'}
              </div>
            )}
          </div>
        </div>
      </div>
    );
  }
}

export default WeatherWidget;
