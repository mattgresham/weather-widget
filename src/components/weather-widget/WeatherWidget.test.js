import React from 'react';
import { render } from 'enzyme';
import toJson from 'enzyme-to-json';
import WeatherWidget from './WeatherWidget';

const mockFn = jest.fn();

it('renders correctly with metric no data', () => {
  const component = render(
    <WeatherWidget
      getWeather={mockFn}
      icon=""
      isWindShown
      location=""
      temperature=""
      title=""
      units="metric"
      wind=""
    />
  );
  expect(toJson(component)).toMatchSnapshot();
});

it('renders correctly with imperial no data', () => {
  const component = render(
    <WeatherWidget
      getWeather={mockFn}
      icon=""
      isWindShown
      location=""
      temperature=""
      title=""
      units="imperial"
      wind=""
    />
  );
  expect(toJson(component)).toMatchSnapshot();
});

it('renders correctly with a title, wind and metric units', () => {
  const component = render(
    <WeatherWidget
      getWeather={mockFn}
      icon="00"
      isWindShown
      location="Australia"
      temperature="10"
      title="Weather"
      units="metric"
      wind="N 10"
    />
  );
  expect(toJson(component)).toMatchSnapshot();
});

it('renders correctly with a title, no wind and metric units', () => {
  const component = render(
    <WeatherWidget
      getWeather={mockFn}
      icon="00"
      isWindShown={false}
      location="Australia"
      temperature="10"
      title="Weather"
      units="metric"
      wind="N 10"
    />
  );
  expect(toJson(component)).toMatchSnapshot();
});
