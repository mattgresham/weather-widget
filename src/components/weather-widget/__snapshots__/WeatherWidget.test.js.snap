// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with a title, no wind and metric units 1`] = `
<div
  class="widget"
>
  <div
    class="title"
  >
    Weather
  </div>
  <div
    class="container"
  >
    <div
      class="iconContainer"
    >
      <img
        alt=""
        class="icon"
        src="https://openweathermap.org/img/w/00.png"
      />
    </div>
    <div
      class="weatherContainer"
    >
      <div>
        Australia
      </div>
      <div
        class="temperature"
      >
        10
        <span
          class="degree"
        >
          째
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders correctly with a title, wind and metric units 1`] = `
<div
  class="widget"
>
  <div
    class="title"
  >
    Weather
  </div>
  <div
    class="container"
  >
    <div
      class="iconContainer"
    >
      <img
        alt=""
        class="icon"
        src="https://openweathermap.org/img/w/00.png"
      />
    </div>
    <div
      class="weatherContainer"
    >
      <div>
        Australia
      </div>
      <div
        class="temperature"
      >
        10
        <span
          class="degree"
        >
          째
        </span>
      </div>
      <div
        class="windContainer"
      >
        <span
          class="windLabel"
        >
          Wind
        </span>
         N 10km/h
      </div>
    </div>
  </div>
</div>
`;

exports[`renders correctly with imperial no data 1`] = `
<div
  class="widget"
>
  <div
    class="title"
  >
    Title of widget
  </div>
  <div
    class="container"
  >
    <div
      class="iconContainer"
    />
    <div
      class="weatherContainer"
    >
      <div>
        -
      </div>
      <div
        class="temperature"
      >
        -
        <span
          class="degree"
        >
          째
        </span>
      </div>
      <div
        class="windContainer"
      >
        <span
          class="windLabel"
        >
          Wind
        </span>
         -mph
      </div>
    </div>
  </div>
</div>
`;

exports[`renders correctly with metric no data 1`] = `
<div
  class="widget"
>
  <div
    class="title"
  >
    Title of widget
  </div>
  <div
    class="container"
  >
    <div
      class="iconContainer"
    />
    <div
      class="weatherContainer"
    >
      <div>
        -
      </div>
      <div
        class="temperature"
      >
        -
        <span
          class="degree"
        >
          째
        </span>
      </div>
      <div
        class="windContainer"
      >
        <span
          class="windLabel"
        >
          Wind
        </span>
         -km/h
      </div>
    </div>
  </div>
</div>
`;
